pex/pex_builder.py

```diff
@@ -253,7 +254,29 @@ def _add_dist_dir(self, path, dist_name):
         self._copy_or_link(filename, target)
     return CacheHelper.dir_hash(path)
 
+  def _get_installer_paths(self, dist_name):
+    """Set up an overrides dict for WheelFile.install that installs the contents
+    of a wheel into its own base in the pex dependencies cache.
+    """
+    base = os.path.join(self.path(), self._pex_info.internal_cache, dist_name)
+    return {
+      'purelib': os.path.join(base),
+      'headers': os.path.join(base, 'headers'),
+      'scripts': os.path.join(base, 'bin'),
+      'platlib': base,
+      'data': base
+    }
+
   def _add_dist_zip(self, path, dist_name):
+    # We need to distinguish between wheels and other zips. Most of the time,
+    # when we have a zip, it contains its contents in an importable form.
+    # But wheels don't have to be importable, so we need to force them
+    # into an importable shape. We can do that by installing it into its own
+    # wheel dir.
+    if dist_name.endswith("whl"):
+      wf = WheelFile(path)
+      wf.install(overrides=self._get_installer_paths(dist_name), force=True)
```

Comment:
seems to me like this conditional block should be mutually exclusive with the zip explosion below.. or else land in an isolated install path?

otherwise you could end up with dirs containing both installed and unzipped wheels, which could be problematic if e.g. inner paths collide?